File: 17-01-25_Product_array_puzzle.py
class Solution:
    def productExceptSelf(self, arr):
        zeros = 0
        idx = -1
        prod = 1
    
        for i in range(len(arr)):
            if arr[i] == 0:
                zeros += 1
                idx = i
            else:
                prod *= arr[i]
    
        res = [0] * len(arr)
    
        if zeros == 0:
            for i in range(len(arr)):
                res[i] = prod // arr[i]
              
        elif zeros == 1:
            res[idx] = prod
    
        return res

File: 17-01-25_Product_array_puzzle.java
class Solution {
    static int[] productExceptSelf(int[] arr) {
        int zeros = 0, idx = -1, prod = 1;
        int n = arr.length;
      
        for (int i = 0; i < n; i++) {
            if (arr[i] == 0) {
                zeros++;
                idx = i;
            } else {
                prod *= arr[i];
            }
        }
      
        int[] res = new int[n];
        Arrays.fill(res, 0);
      
        if (zeros == 0) {
            for (int i = 0; i < n; i++) 
                res[i] = prod / arr[i];
        } else if (zeros == 1) {
          res[idx] = prod;
        }
      
        return res;
    }
}
